internal/kubeclient/indexer.go

```diff
@@ -498,100 +496,48 @@ func indexStagesByWarehouse(obj client.Object) []string {
-// OIDC email annotations.
+// IndexServiceAccountsByOIDCClaim sets up indexing of ServiceAccounts by
+// their OIDC claim annotations.
 //
-// It configures the cluster's field indexer to allow querying ServiceAccounts
-// using the ServiceAccountsByOIDCEmailIndexField selector.
-func IndexServiceAccountsByOIDCEmail(ctx context.Context, clstr cluster.Cluster) error {
-	return clstr.GetFieldIndexer().IndexField(
-		ctx,
-		&corev1.ServiceAccount{},
-		ServiceAccountsByOIDCEmailIndexField,
-		indexServiceAccountsOIDCEmail,
-	)
-}
-
-// indexServiceAccountsOIDCEmail is a client.IndexerFunc that indexes
-// ServiceAccounts by their OIDC email annotations.
-func indexServiceAccountsOIDCEmail(obj client.Object) []string {
-	sa := obj.(*corev1.ServiceAccount) // nolint: forcetypeassert
-	rawEmails := strings.TrimSpace(sa.GetAnnotations()[rbacapi.AnnotationKeyOIDCEmails])
-	if rawEmails == "" {
-		return nil
-	}
-	emails := strings.Split(rawEmails, ",")
-	refinedEmails := make([]string, 0, len(emails))
-	for _, e := range emails {
-		if email := strings.TrimSpace(e); email != "" {
-			refinedEmails = append(refinedEmails, email)
-		}
-	}
-	return refinedEmails
-}
-
-// IndexServiceAccountsByOIDCGroups sets up indexing of ServiceAccounts by
-// their OIDC group annotations.
-//
-// It configures the cluster's field indexer to allow querying ServiceAccounts
-// using the ServiceAccountsByOIDCGroupIndexField selector.
-func IndexServiceAccountsByOIDCGroups(ctx context.Context, clstr cluster.Cluster) error {
-	return clstr.GetFieldIndexer().IndexField(
+// It configures the manager's field indexer to allow querying ServiceAccounts
+// using the ServiceAccountsByOIDCClaimIndexField selector.
+func IndexServiceAccountsByOIDCClaim(ctx context.Context, mgr ctrl.Manager) error {
+	return mgr.GetFieldIndexer().IndexField(
 		ctx,
 		&corev1.ServiceAccount{},
-		ServiceAccountsByOIDCGroupIndexField,
-		indexServiceAccountsByOIDCGroups,
+		ServiceAccountsByOIDCClaimIndexField,
+		indexServiceAccountsOIDCClaim,
 	)
 }
 
-// indexServiceAccountsByOIDCGroups is a client.IndexerFunc that indexes
-// ServiceAccounts by their OIDC group annotations.
-func indexServiceAccountsByOIDCGroups(obj client.Object) []string {
+// indexServiceAccountsOIDCClaim is a client.IndexerFunc that indexes ServiceAccounts by the
+// OIDC claim name and value annotations.
+func indexServiceAccountsOIDCClaim(obj client.Object) []string {
 	sa := obj.(*corev1.ServiceAccount) // nolint: forcetypeassert
-	rawGroups := strings.TrimSpace(sa.GetAnnotations()[rbacapi.AnnotationKeyOIDCGroups])
-	if rawGroups == "" {
-		return nil
-	}
-	groups := strings.Split(rawGroups, ",")
-	refinedGroups := make([]string, 0, len(groups))
-	for _, g := range groups {
-		if group := strings.TrimSpace(g); group != "" {
-			refinedGroups = append(refinedGroups, group)
+	rawClaimValues := map[string]string{}
+	for annotationKey, annotationValue := range sa.GetAnnotations() {
```

Comment:
I know you got this pattern from some of the existing code that you replaced, but it was inefficient and if you wouldn't mind improving it by combining the two for loops in this function, I'd be grateful for that.

Reply:
Yes, that makes sense, didn't even question it when I did it, I've combined the loops which hopefully makes more sense now.